File: sdk/wauth-py/src/wauth_sdk/profiles.py
from typing import Any, Dict, List


def _as_dict(value: Any) -> Dict[str, Any]:
    if not isinstance(value, dict):
        raise ValueError("Expected object")
    return value


def evaluate_requester_continuity(value: Any) -> str:
    data = _as_dict(value)
    return "allow" if data.get("continuity") == "intact" else "step_up_or_deny"


def evaluate_instruction_source(value: Any) -> str:
    data = _as_dict(value)
    if data.get("mutability") == "externally_editable" or data.get("review_state") != "reviewed":
        return "data"
    return "instruction"


def evaluate_execution_budget(value: Any) -> str:
    data = _as_dict(value)
    persistent_types = {"cron", "daemon", "monitor", "recurring_workflow"}
    if not data.get("may_create_persistent_process", True) and data.get("requested_process_type") in persistent_types:
        return "deny"
    return "allow"


def evaluate_postcondition(value: Any) -> str:
    data = _as_dict(value)
    claimed = data.get("claimed_success") is True
    verified = data.get("verified_success") is True
    if claimed and not verified:
        return "unverified_or_failed"
    if verified:
        return "verified"
    return "unknown"


def validate_provenance_chain(value: Any) -> bool:
    data = _as_dict(value)
    events = data.get("events")
    if not isinstance(events, list) or not events:
        return False

    for idx, raw_event in enumerate(events):
        event = _as_dict(raw_event)
        if idx == 0:
            if event.get("prev_event_hash") is not None:
                return False
            continue
        previous = _as_dict(events[idx - 1])
        if event.get("prev_event_hash") != previous.get("event_hash"):
            return False

    return True


def evaluate_risk_policy(value: Any) -> str:
    data = _as_dict(value)
    risk = _as_dict(data.get("risk", {}))
    return "wauth_required" if risk.get("recommended_action") == "step_up" else "permit"


def evaluate_multi_agent_trust(value: Any) -> str:
    data = _as_dict(value)
    independent = data.get("verified_independently") is True
    if not independent and data.get("assertion_basis") == "shared_channel":
        return "insufficient_without_external_anchor"
    return "sufficient"

File: sdk/wauth-ts/src/profiles.ts
import type { JsonValue } from "./types.js";

function asRecord(value: JsonValue): Record<string, JsonValue> {
  if (typeof value === "object" && value !== null && !Array.isArray(value)) {
    return value;
  }
  throw new Error("Expected object");
}

export function evaluateRequesterContinuity(input: JsonValue): string {
  const record = asRecord(input);
  return record.continuity === "intact" ? "allow" : "step_up_or_deny";
}

export function evaluateInstructionSource(input: JsonValue): string {
  const record = asRecord(input);
  const mutability = record.mutability;
  const reviewState = record.review_state;
  if (mutability === "externally_editable" || reviewState !== "reviewed") {
    return "data";
  }
  return "instruction";
}

export function evaluateExecutionBudget(input: JsonValue): string {
  const record = asRecord(input);
  const mayPersist = record.may_create_persistent_process;
  const processType = record.requested_process_type;
  const persistentTypes = new Set(["cron", "daemon", "monitor", "recurring_workflow"]);
  if (mayPersist === false && typeof processType === "string" && persistentTypes.has(processType)) {
    return "deny";
  }
  return "allow";
}

export function evaluatePostcondition(input: JsonValue): string {
  const record = asRecord(input);
  const claimed = record.claimed_success === true;
  const verified = record.verified_success === true;
  if (claimed && !verified) {
    return "unverified_or_failed";
  }
  return verified ? "verified" : "unknown";
}

export function validateProvenanceChain(input: JsonValue): boolean {
  const record = asRecord(input);
  const events = record.events;
  if (!Array.isArray(events) || events.length === 0) {
    return false;
  }

  for (let i = 0; i < events.length; i += 1) {
    const event = asRecord(events[i] as JsonValue);
    const prev = event.prev_event_hash;
    if (i === 0) {
      if (prev !== null) {
        return false;
      }
      continue;
    }

    const previous = asRecord(events[i - 1] as JsonValue);
    if (prev !== previous.event_hash) {
      return false;
    }
  }

  return true;
}

export function evaluateRiskPolicy(input: JsonValue): string {
  const record = asRecord(input);
  const risk = asRecord(record.risk as JsonValue);
  return risk.recommended_action === "step_up" ? "wauth_required" : "permit";
}

export function evaluateMultiAgentTrust(input: JsonValue): string {
  const record = asRecord(input);
  const basis = record.assertion_basis;
  const independent = record.verified_independently === true;
  if (!independent && basis === "shared_channel") {
    return "insufficient_without_external_anchor";
  }
  return "sufficient";
}
